File: Programmers/2022_kakao_internship/5/main.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<vector<int>> ShiftRow(vector<vector<int>> rc, int n, int m) {
    vector<vector<int>> temp;
    for (int i=0; i<n; i++) {
        vector<int> line;
        for (int j=0; j<m; j++) {
            line.push_back(rc[(i-1+n)%n][j]);
        }
        temp.push_back(line);
    }
    return temp;
}

vector<vector<int>> Rotate(vector<vector<int>> rc, int n, int m) {
    vector<vector<int>> temp; 
    for (int i=0; i<n; i++) {
        vector<int> line;
        for (int j=0; j<m; j++) {
            line.push_back(rc[i][j]);
        }
        temp.push_back(line);
    }

    // for (int i=0; i<n; i++) {
    //     for (int j=0; j<m; j++) {
    //         cout << temp[i][j] << " ";
    //     }
    //     cout << endl;
    // }

    vector<int> line;
    for (int i=0; i<m-1; i++) line.push_back(rc[0][i]);
    for (int i=0; i<n-1; i++) line.push_back(rc[i][m-1]);
    for (int i=m-1; i>0; i--) line.push_back(rc[n-1][i]);
    for (int i=n-1; i>0; i--) line.push_back(rc[i][0]);

    int idx = 0;
    temp[0][0] = line.back();
    for (int i=1; i<m-1; i++) {
        temp[0][i] = line.front();
        line.erase(line.begin());
    }
    for (int i=0; i<n-1; i++) {
        temp[i][m-1] = line.front();
        line.erase(line.begin());
    }
    for (int i=m-1; i>0; i--) {
        temp[n-1][i] = line.front();
        line.erase(line.begin());
    }
    for (int i=n-1; i>0; i--) {
        temp[i][0] = line.front();
        line.erase(line.begin());
    }
    
    return temp;
}

vector<vector<int>> solution(vector<vector<int>> rc, vector<string> operations) {
    vector<vector<int>> answer;
    int n = rc.size();
    int m = rc[0].size();
    for (int i=0; i<operations.size(); i++) {
        if (operations[i].compare("ShiftRow") == 0) {
            answer = ShiftRow(rc, n, m);
            rc = answer;

            for (int i=0; i<rc.size(); i++) {
                for (int j=0; j<rc[0].size(); j++) {
                    cout << answer[i][j] << " ";
                }
                cout << endl;
            } cout << endl;

        }
        else {
            answer = Rotate(rc, n, m);
            rc = answer;

            for (int i=0; i<rc.size(); i++) {
                for (int j=0; j<rc[0].size(); j++) {
                    cout << answer[i][j] << " ";
                }
                cout << endl;
            } cout << endl;
        }
    }
    
    return answer;
}

int main() {
    // vector<vector<int>> rc = {{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}};
    // vector<string> operations = {"ShiftRow", "Rotate", "ShiftRow", "Rotate"};
    vector<vector<int>> rc = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    vector<string> operations = {"Rotate", "ShiftRow"};
    vector<vector<int>> answer = solution(rc, operations);
    for (int i=0; i<rc.size(); i++) {
        for (int j=0; j<rc[0].size(); j++) {
            cout << answer[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Programmers/2022_kakao_internship/5/main.py
num2loc, max_idx = dict(), 0 # num2loc = {인덱스:위치}

def MakeDict(n, m):
    global num2loc, max_idx

    for i in range(m-1):
        num2loc[max_idx] = (0, i)
        max_idx += 1
    for i in range(n-1):
        num2loc[max_idx] = (i, m-1)
        max_idx += 1
    for i in range(m-1, 0, -1):
        num2loc[max_idx] = (n-1, i)
        max_idx += 1
    for i in range(n-1, 0, -1):
        num2loc[max_idx] = (i, 0)
        max_idx += 1
    max_idx -= 1

def ShiftRow(rc):
    temp = [[] for _ in range(len(rc))]
    for i, row in enumerate(rc):
        if i == len(rc)-1:
            temp[0] = row
            break
        temp[i+1] = row
    return temp

def Rotate(rc):
    global max_idx
    temp = [[c for c in r] for r in rc]
    for idx, loc in num2loc.items():
        x, y = loc
        nx, ny = 0, 0
        if idx < max_idx:
            nx, ny = num2loc[idx+1]
        temp[nx][ny] = rc[x][y]
    return temp

def solution(rc, operations):
    n, m = len(rc), len(rc[0])
    MakeDict(n, m)

    for operation in operations:
        if operation == "Rotate": rc = Rotate(rc)
        else: rc = ShiftRow(rc)
    return rc

print(solution([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]], ["ShiftRow", "Rotate", "ShiftRow", "Rotate"]))
print(solution([[1, 2, 3], [4, 5, 6], [7, 8, 9]], ["Rotate", "ShiftRow"]))